backend/src/apiserver/resource/resource_manager.go

```diff
@@ -195,616 +510,1147 @@ func (r *ResourceManager) ArchiveExperiment(ctx context.Context, experimentId st
+		return util.Wrapf(err, "Failed to delete a DB entry for pipeline version id %v", pipelineVersionId)
 	}
 	return nil
 }
 
-func (r *ResourceManager) CreateTask(ctx context.Context, apiTask *apiv1beta1.Task) (*model.Task, error) {
-	uuid, err := r.uuid.NewRandom()
+// Creates a run and schedule a workflow CR.
+// Note: when creating a run from a manifest, this triggers creation of
+// a new pipeline and pipeline version that share the name, description, and namespace.
+// If run.ExperimentId is not specified, it is set to the default experiment.
+// If run.Namespace  is no specified, it gets inferred from the parent experiment.
+// Manifest's namespace gets overwritten with the run.Namespace.
+// Creating a run from recurring run prioritizes recurring run's pipeline spec over the run's one.
+func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model.Run, error) {
+	if run.ExperimentId == "" {
+		defaultExperimentId, err := r.GetDefaultExperimentId()
+		if err != nil {
+			return nil, util.Wrapf(err, "Failed to create a run with empty experiment id. Specify experiment id for the run or check if the default experiment exists")
+		}
+		run.ExperimentId = defaultExperimentId
+	}
+
+	// Validate namespace
+	if run.Namespace == "" {
+		namespace, err := r.GetNamespaceFromExperimentId(run.ExperimentId)
+		if err != nil {
+			return nil, util.Wrap(err, "Failed to create a run")
+		}
+		run.Namespace = namespace
+	}
+	if common.IsMultiUserMode() {
+		if run.Namespace == "" {
+			return nil, util.NewInternalServerError(util.NewInvalidInputError("Run cannot have an empty namespace in multi-user mode"), "Failed to create a run")
+		}
+	}
+	if err := r.ValidateExperimentNamespace(run.ExperimentId, run.Namespace); err != nil {
+		return nil, util.Wrap(err, "Failed to create a run")
+	}
+
+	// Validate pipeline version id
+	var pipelineVersion *model.PipelineVersion
+	if run.PipelineSpec.PipelineVersionId != "" {
+		pv, err := r.GetPipelineVersion(run.PipelineSpec.PipelineVersionId)
+		if err != nil {
+			return nil, util.Wrapf(err, "Failed to create a run with an invalid pipeline version id: %v", run.PipelineSpec.PipelineVersionId)
+		}
+		pipelineVersion = pv
+	} else if run.PipelineSpec.PipelineId != "" {
+		pv, err := r.GetLatestPipelineVersion(run.PipelineSpec.PipelineId)
+		if err != nil {
+			return nil, util.Wrap(err, "Failed to create a run with an empty pipeline version id")
+		}
+		pipelineVersion = pv
+	} else if run.PipelineSpec.PipelineSpecManifest != "" || run.PipelineSpec.WorkflowSpecManifest != "" {
+		manifest := run.PipelineSpec.PipelineSpecManifest
+		if manifest == "" {
+			manifest = run.PipelineSpec.WorkflowSpecManifest
+		}
+		pipeline := &model.Pipeline{
+			Name:        run.DisplayName,
+			Description: run.Description,
+			Namespace:   run.Namespace,
+		}
+		pipeline, err := r.CreatePipeline(pipeline)
```

Comment:
From the logic here, if a run does not specify pipeline id or pipeline version id, but provide the workflow manifest/pipeline manifest, the code will create a new pipeline for it. This is a behavior change I'm not sure if we want, especially in this PR.

Reply:
This is correct. Submitting a new pipeline with a spec leads to creation of a new pipeline and a new pipeline version under it. This is necessary to support the old behavior without changes for the users. Note, `Pipeline` does not contain pipeline spec in KFP v2.